jupyter_server/serverapp.py

```diff
@@ -1907,42 +1991,40 @@ def remove_browser_open_file(self):
             if e.errno != errno.ENOENT:
                 raise
 
-    def launch_browser(self):
-        try:
-            browser = webbrowser.get(self.browser or None)
-        except webbrowser.Error as e:
-            self.log.warning(_('No web browser found: %s.') % e)
-            browser = None
-
-        if not browser:
-            return
-
+    def _prepare_browser_open(self):
         if not self.use_redirect_file:
             uri = self.default_url[len(self.base_url):]
 
             if self.token:
                 uri = url_concat(uri, {'token': self.token})
 
         if self.file_to_run:
-            if not os.path.exists(self.file_to_run):
-                self.log.critical(_("%s does not exist") % self.file_to_run)
-                self.exit(1)
-
-            relpath = os.path.relpath(self.file_to_run, self.root_dir)
-            uri = url_escape(url_path_join('notebooks', *relpath.split(os.sep)))
-
-            # Write a temporary file to open in the browser
-            fd, open_file = tempfile.mkstemp(suffix='.html')
-            with open(fd, 'w', encoding='utf-8') as fh:
-                self._write_browser_open_file(uri, fh)
+            # Create a separate, temporary open-browser-file
+            # pointing at a specific file.
+            open_file = self.browser_open_file_to_run
         else:
+            # otherwise, just return the usual open browser file.
             open_file = self.browser_open_file
 
         if self.use_redirect_file:
             assembled_url = urljoin('file:', pathname2url(open_file))
         else:
             assembled_url = url_path_join(self.connection_url, uri)
 
+        return assembled_url, open_file
+
+    def launch_browser(self):
+        try:
+            browser = webbrowser.get(self.browser or None)
+        except webbrowser.Error as e:
+            self.log.warning(_('No web browser found: %s.') % e)
```

Comment:
Hi,

When I start jupyter from CUI, the program will stop with the following error.
```bash
...
  File "/root/.pyenv/versions/3.7.3/lib/python3.7/site-packages/jupyter_server/serverapp.py", line 2020, in launch_browser
    self.log.warning(_('No web browser found: %s.') % e)
UnboundLocalError: local variable '_' referenced before assignment
```

As shown in the error message, by removing the `_` from the relevant part, jupyter can be executed successfully.

Looking at this script, it seems that underscores `_` are inserted in `self.log.info()` and so on. 
I am not familiar with this notation, so I would like to confirm it.

Thanks in advance,

Reply:
Notebook server and Jupyter Server (due to its derivation from the former) [define `_` to be the method `trans.gettext`](https://github.com/afshin/jupyter_server/blob/file-to-run/jupyter_server/transutils.py#L13) such that any strings wrapped in `_()` are candidates for translation lookups.  However, this precludes the common use of `_` to represent an ignored value in the same method - as is the case [here](https://github.com/jupyter-server/jupyter_server/blob/master/jupyter_server/serverapp.py#L2026) (a few lines below):
```python
        assembled_url, _ = self._prepare_browser_open()
```

Seems like we should no longer return `open_file` from `_prepare_browser_open()` (since this is the only call to it) or adjust the name for the ignored value.

I don't know the history of using `_` for `trans.gettext` but using something like `_i18n` instead seems more clear while not being too intrusive.

Reply:
Thanks for the clarification!

Now I see how `_` is used here, but as it is pointed out, it is neither clear nor safe usage, I guess...
This is a common way to represent ignored variables with `_` in python, so I agree with @kevin-bates's suggestion!

Reply:
I've created #424 to help avoid this in the future.